Add documentation for InputChanges (#8940)
gradle/gradle: subprojects/docs/src/docs/userguide/custom_tasks.adoc

[[incremental_tasks]]
== Incremental tasks

With Gradle, it's very simple to implement a task that is skipped when all of its inputs and outputs are up to date (see <<more_about_tasks.adoc#sec:up_to_date_checks,Incremental Builds>>).
However, there are times when only a few input files have changed since the last execution, and you'd like to avoid reprocessing all of the unchanged inputs.

> This can be particularly useful for a transformer task, that converts input files to output files on a 1:1 basis.

Reviewer: No comma. "...for a transformer task that converts..."

Author: Removed the comma in https://github.com/gradle/gradle/pull/8952.

> If you'd like to optimize your build so that only out-of-date input files are reprocessed, you can do so with an _incremental task_.

Reviewer: I don't understand why you want to use "reprocess" rather than "process". I think I just don't like the word "reprocess". I think if the same files were processed more than once during the execution of a task, then "reprocess" might be appropriate. Is that the case here?

Author: Processed it is. See https://github.com/gradle/gradle/pull/8952.

[NOTE]
====
There is the link:{javadocPath}/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html[IncrementalTaskInputs] API, which is available in Gradle versions before 5.4.
The usage of this API is discouraged, since it does not allow querying individual input properties for changes.

> The `IncrementalTaskInputs` API will be deprecated an eventually removed.

Reviewer: "an" -> "and"

Author: Done in https://github.com/gradle/gradle/pull/8952.

> If you need to use the old API, have a look at the documentation in the link:https://docs.gradle.org/5.3.1/userguide/custom_tasks.html#incremental_tasks[userguide for Gradle 5.3.1].

Reviewer: "userguide" -> "user manual".

Author: Done in https://github.com/gradle/gradle/pull/8952.

With Gradle, it's very simple to implement a task that is skipped when all of its inputs and outputs are up to date (see <<more_about_tasks.adoc#sec:up_to_date_checks,Incremental Builds>>). However, there are times when only a few input files have changed since the last execution, and you'd like to avoid reprocessing all of the unchanged inputs. This can be particularly useful for a transformer task, that converts input files to output files on a 1:1 basis.

> If you'd like to optimize your build so that only out-of-date inputs are processed, you can do so with an _incremental task_.

Reviewer 2: ```suggestion
If you'd like to optimize your build so that only out-of-date input files are reprocessed, you can do so with an _incremental task_.
```

[NOTE]
====

> There is the link:{javadocPath}/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html[IncrementalTaskInputs] API, which is available in Gradle versions smaller than 5.4.

Reviewer 2: ```suggestion
There is the link:{javadocPath}/org/gradle/api/tasks/incremental/IncrementalTaskInputs.html[IncrementalTaskInputs] API, which is available in Gradle versions before 5.4.
```

> The usage of this API is discouraged, since it does not allow querying individual input properties for changes.

Reviewer 2: Should we mention that it will eventually be deprecated and removed?

Author: I added a note.

If you need to use the old API, have a look at the documentation in the link:https://docs.gradle.org/5.3.1/userguide/custom_tasks.html#incremental_tasks[userguide for Gradle 5.3.1].
====

[[sec:implementing_an_incremental_task]]
=== Implementing an incremental task

For a task to process inputs incrementally, that task must contain an _incremental task action_.
This is a task action method that contains a single link:{groovyDslPath}/org.gradle.work.InputChanges.html[InputChanges] parameter, which indicates to Gradle that the action will process the changed inputs only.
Moreover, the task needs to declare at least one incremental file input property, by using either `@link:{javadocPath}/org/gradle/work/Incremental.html[Incremental]` or `@link:{javadocPath}/org/gradle/api/tasks/SkipWhenEmpty.html[SkipWhenEmpty]`.

[IMPORTANT]
====

> For being able to query incremental changes for an input property, the getter of that property always needs to return the same value.

Reviewer 2: ```suggestion
To query incremental changes for an input file property, the getter of that property always needs to return the same instance.
```

The easiest way to accomplish this is to use `link:{javadocPath}/org/gradle/api/file/RegularFileProperty.html[RegularFileProperty]`, `link:{javadocPath}/org/gradle/api/file/DirectoryProperty.html[DirectoryProperty]` or `link:{javadocPath}/org/gradle/api/file/ConfigurableFileCollection.html[ConfigurableFileCollection]` as type of the property.

> See the chapter on <<lazy_configuration.adoc#lazy_configuration,lazy configuration>>, especially the sections about <<lazy_configuration.adoc#sec:lazy_properties,using read-only and configurable properties>> and <<lazy_configuration.adoc#sec:working_with_files_in_lazy_properties,lazy file properties>>.

Reviewer 2: ```suggestion
See also the chapter on <<lazy_configuration.adoc#lazy_configuration,lazy configuration>>, especially the sections about <<lazy_configuration.adoc#sec:lazy_properties,using read-only and configurable properties>> and <<lazy_configuration.adoc#sec:working_with_files_in_lazy_properties,lazy file properties>>.
```

[[sec:which_inputs_are_considered_out_of_date]]
=== Which inputs are considered out of date?

> When Gradle has history of a previous task execution, and the only changes to the task execution context since that execution are to incremental input file properties, then Gradle is able to determine which input files need to be reprocessed by the task.

Reviewer 2: Do we define what the "task execution context" is? Can we use a better term perhaps?

Author: How about we say that `and the only changes to the task inputs and outputs since` instead?

Author: I reworded the sentence.

In this case, the link:{groovyDslPath}/org.gradle.work.InputChanges.html#org.gradle.work.InputChanges:getFileChanges(org.gradle.api.file.FileCollection)[InputChanges.getFileChanges] method returns details for all input files for the given property that were _added_, _modified_ or _removed_.

However, there are many cases where Gradle is unable to determine which input files need to be reprocessed. Examples include:

* There is no history available from a previous execution.
* You are building with a different version of Gradle. Currently, Gradle does not use task history from a different version.
* An `upToDateWhen` criteria added to the task returns `false`.
* An input property has changed since the previous execution.
* A non-incremental input file property has changed since the previous execution.
* One or more output files have changed since the previous execution.

> In any of these cases, Gradle will consider all of the input files to be `added`.

Reviewer 2: ```suggestion
In any of these cases, Gradle will report all input files as `ADDED`.
```

> include::{samplesPath}/userguide/tasks/incrementalTask/incrementalTaskNoChange.out[]

Reviewer: I know this isn't part of the change, but this file is empty. Can we not have it print _something_ out? It just looks like a broken example right now.

Reviewer: @wolfs :point_up: (in case you don't get a notification from the above comment)

Author: I guess if we run the build with `--verbose` we will get the task headers, right?

[[sec:storing_incremental_task_state]]
=== Storing incremental state for cached tasks

> Using Gradle's `InputChanges` is not the only way to create tasks that only works on changes since the last execution.

Reviewer 2: ```suggestion
Using Gradle's `InputChanges` is not the only way to create tasks that only work on changes since the last execution.
```

[[sec:storing_incremental_task_state]]
=== Storing incremental state for cached tasks

Using Gradle's `InputChanges` is not the only way to create tasks that only work on changes since the last execution.
Tools like the Kotlin compiler provide incrementality as a built-in feature.
The way this is typically implemented is that the tool stores some analysis data about the state of the previous execution in some file.
If such state files are <<build_cache.adoc#sec:task_output_caching_inputs,relocatable>>, then they can be declared as outputs of the task.
This way when the task's results are loaded from cache, the next execution can already use the analysis data loaded from cache, too.

However, if the state files are non-relocatable, then they can't be shared via the build cache.
Indeed, when the task is loaded from cache, any such state files must be cleaned up to prevent stale state to confuse the tool during the next execution.

> Gradle can ensure such stale files are removed if they are declared via `task.localState.register()` or a property is marked with the `@LocalState` annotation.

Reviewer: @wolfs Is there any API documentation for `register()` that we can link to?

Author: That is the method: https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskLocalState.html#register-java.lang.Object...-

I guess we should expose the `localState` property on the `Task` interface in the DSL doc as well.

Author: Added the link in https://github.com/gradle/gradle/pull/8983.
